### ISSUE TITLE
MacOS/Sonoma: Crash when building `odbc` from source

### ISSUE THREAD

**Issue by detule** [r-dbi/odbc#624]
[Other users](https://github.com/r-dbi/odbc/issues/624#issue-2010813484) have [reported this as well](https://github.com/r-dbi/odbc/issues/611), but I am running into this for the first time.

```
> install.packages("odbc", type = "source")
trying URL 'https://mirror.las.iastate.edu/CRAN/src/contrib/odbc_1.3.5.tar.gz'
Content type 'application/x-gzip' length 348241 bytes (340 KB)
==================================================
downloaded 340 KB

* installing *source* package ‘odbc’ ...
** package ‘odbc’ successfully unpacked and MD5 sums checked
** using staged installation
Found odbc_config cflags and libs!
PKG_CFLAGS=-DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include
PKG_LIBS=-L/opt/R/x86_64/lib -lodbc
** libs
using C++ compiler: ‘Apple clang version 15.0.0 (clang-1500.0.40.1)’
using SDK: ‘MacOSX14.0.sdk’
(cd cctz && \
make libcctz.a CC="clang -arch x86_64" CXX="clang++ -arch x86_64 -std=gnu++17" CXXFLAGS="-falign-functions=64 -Wall -g -O2 " AR="ar" ARFLAGS=rv CXXPICFLAGS="-fPIC")
clang++ -arch x86_64 -std=gnu++17 -falign-functions=64 -Wall -g -O2  -Wall -Iinclude -std=c++11 -pthread  -O -fPIC -MMD  -c -o time_zone_format.o src/time_zone_format.cc
clang++ -arch x86_64 -std=gnu++17 -falign-functions=64 -Wall -g -O2  -Wall -Iinclude -std=c++11 -pthread  -O -fPIC -MMD  -c -o time_zone_if.o src/time_zone_if.cc
clang++ -arch x86_64 -std=gnu++17 -falign-functions=64 -Wall -g -O2  -Wall -Iinclude -std=c++11 -pthread  -O -fPIC -MMD  -c -o time_zone_impl.o src/time_zone_impl.cc
clang++ -arch x86_64 -std=gnu++17 -falign-functions=64 -Wall -g -O2  -Wall -Iinclude -std=c++11 -pthread  -O -fPIC -MMD  -c -o time_zone_info.o src/time_zone_info.cc
clang++ -arch x86_64 -std=gnu++17 -falign-functions=64 -Wall -g -O2  -Wall -Iinclude -std=c++11 -pthread  -O -fPIC -MMD  -c -o time_zone_libc.o src/time_zone_libc.cc
clang++ -arch x86_64 -std=gnu++17 -falign-functions=64 -Wall -g -O2  -Wall -Iinclude -std=c++11 -pthread  -O -fPIC -MMD  -c -o time_zone_lookup.o src/time_zone_lookup.cc
clang++ -arch x86_64 -std=gnu++17 -falign-functions=64 -Wall -g -O2  -Wall -Iinclude -std=c++11 -pthread  -O -fPIC -MMD  -c -o time_zone_posix.o src/time_zone_posix.cc
ar rv libcctz.a time_zone_format.o time_zone_if.o time_zone_impl.o time_zone_info.o time_zone_libc.o time_zone_lookup.o time_zone_posix.o
ar: creating archive libcctz.a
a - time_zone_format.o
a - time_zone_if.o
a - time_zone_impl.o
a - time_zone_info.o
a - time_zone_libc.o
a - time_zone_lookup.o
a - time_zone_posix.o
clang++ -arch x86_64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include  -I'/Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include' -I/opt/R/x86_64/include   -Icctz/include -Inanodbc -I. -DBUILD_REAL_64_BIT_MODE -DNANODBC_ODBC_VERSION=SQL_OV_ODBC3 -fPIC -fPIC  -falign-functions=64 -Wall -g -O2  -c odbc_result.cpp -o odbc_result.o
clang++ -arch x86_64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include  -I'/Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include' -I/opt/R/x86_64/include   -Icctz/include -Inanodbc -I. -DBUILD_REAL_64_BIT_MODE -DNANODBC_ODBC_VERSION=SQL_OV_ODBC3 -fPIC -fPIC  -falign-functions=64 -Wall -g -O2  -c connection.cpp -o connection.o
clang++ -arch x86_64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include  -I'/Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include' -I/opt/R/x86_64/include   -Icctz/include -Inanodbc -I. -DBUILD_REAL_64_BIT_MODE -DNANODBC_ODBC_VERSION=SQL_OV_ODBC3 -fPIC -fPIC  -falign-functions=64 -Wall -g -O2  -DNANODBC_OVERALLOCATE_CHAR -c nanodbc/nanodbc.cpp -o nanodbc.o
nanodbc/nanodbc.cpp:301:32: warning: 'codecvt_utf8_utf16<char16_t>' is deprecated [-Wdeprecated-declarations]
    out = std::wstring_convert<NANODBC_CODECVT_TYPE<wide_char_t>, wide_char_t>().to_bytes(in);
                               ^
nanodbc/nanodbc.cpp:144:35: note: expanded from macro 'NANODBC_CODECVT_TYPE'
#define NANODBC_CODECVT_TYPE std::codecvt_utf8_utf16
                                  ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/codecvt:541:28: note: 'codecvt_utf8_utf16<char16_t>' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 codecvt_utf8_utf16
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:862:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:835:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
nanodbc/nanodbc.cpp:301:16: warning: 'wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>' is deprecated [-Wdeprecated-declarations]
    out = std::wstring_convert<NANODBC_CODECVT_TYPE<wide_char_t>, wide_char_t>().to_bytes(in);
               ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/locale:3603:28: note: 'wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 wstring_convert
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:862:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:835:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
2 warnings generated.
clang++ -arch x86_64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include  -I'/Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include' -I/opt/R/x86_64/include   -Icctz/include -Inanodbc -I. -DBUILD_REAL_64_BIT_MODE -DNANODBC_ODBC_VERSION=SQL_OV_ODBC3 -fPIC -fPIC  -falign-functions=64 -Wall -g -O2  -c result.cpp -o result.o
clang++ -arch x86_64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include  -I'/Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include' -I/opt/R/x86_64/include   -Icctz/include -Inanodbc -I. -DBUILD_REAL_64_BIT_MODE -DNANODBC_ODBC_VERSION=SQL_OV_ODBC3 -fPIC -fPIC  -falign-functions=64 -Wall -g -O2  -c odbc_connection.cpp -o odbc_connection.o
clang++ -arch x86_64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include  -I'/Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include' -I/opt/R/x86_64/include   -Icctz/include -Inanodbc -I. -DBUILD_REAL_64_BIT_MODE -DNANODBC_ODBC_VERSION=SQL_OV_ODBC3 -fPIC -fPIC  -falign-functions=64 -Wall -g -O2  -c RcppExports.cpp -o RcppExports.o
clang++ -arch x86_64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include  -I'/Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include' -I/opt/R/x86_64/include   -Icctz/include -Inanodbc -I. -DBUILD_REAL_64_BIT_MODE -DNANODBC_ODBC_VERSION=SQL_OV_ODBC3 -fPIC -fPIC  -falign-functions=64 -Wall -g -O2  -c Iconv.cpp -o Iconv.o
clang++ -arch x86_64 -std=gnu++17 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include  -I'/Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include' -I/opt/R/x86_64/include   -Icctz/include -Inanodbc -I. -DBUILD_REAL_64_BIT_MODE -DNANODBC_ODBC_VERSION=SQL_OV_ODBC3 -fPIC -fPIC  -falign-functions=64 -Wall -g -O2  -c utils.cpp -o utils.o
clang++ -arch x86_64 -std=gnu++17 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -L/Library/Frameworks/R.framework/Resources/lib -L/opt/R/x86_64/lib -o odbc.so odbc_result.o connection.o nanodbc.o result.o odbc_connection.o RcppExports.o Iconv.o utils.o -L/opt/R/x86_64/lib -lodbc -Lcctz -lcctz -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
installing to /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/00LOCK-odbc/00new/odbc/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (odbc)

The downloaded source packages are in
	‘/private/var/folders/d6/p5whp6q52njchl8r8p29hnkh0000gp/T/RtmpDAPZKX/downloaded_packages’
> library(odbc)
> conn <- DBI::dbConnect(odbc::odbc(), dsn="snowflake_db", uid="<snip>", pwd="<pwd>")

 *** caught segfault ***
address 0x2202, cause 'memory not mapped'

Traceback:
 1: connection_info(ptr)
 2: OdbcConnection(dsn = dsn, ..., timezone = timezone, timezone_out = timezone_out,     encoding = encoding, bigint = bigint, timeout = timeout,     driver = driver, server = server, database = database, uid = uid,     pwd = pwd, dbms.name = dbms.name, attributes = attributes,     .connection_string = .connection_string)
 3: .local(drv, ...)
 4: DBI::dbConnect(odbc::odbc(), dsn = "snowflake_db", uid = "sa",     pwd = "Yukon900")
 5: DBI::dbConnect(odbc::odbc(), dsn = "snowflake_db", uid = "sa",     pwd = "Yukon900")

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
```
```
R version 4.3.2 (2023-10-31) -- "Eye Holes"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin20 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> sessionInfo()
R version 4.3.2 (2023-10-31)
Platform: x86_64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.0

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
[1] compiler_4.3.2
```

```

**Comment by detule** [comment]
Simon,

As another data-point:

If you take a look at the CRAN install log for the package: https://www.r-project.org/nosvn/R.check/r-release-macos-x86_64/odbc-00install.html you can see numerous deprecation warnings for essentially all of the ODBC API.  I think these deprecation warnings are (only) part of the SQL headers that ship with `iODBC`.

Appreciate the amount of time you have invested in this, hope we can work together to see this through.

**Comment by detule** [comment]
Downgrading to (Homebrew) clang 14 ( from Apple 15 ) makes the issue disappear

```
> install.packages("odbc", type = "source")
--- Please select a CRAN mirror for use in this session ---
trying URL 'https://repo.miserver.it.umich.edu/cran/src/contrib/odbc_1.3.5.tar.gz'
Content type 'application/octet-stream' length 348241 bytes (340 KB)
==================================================
downloaded 340 KB

* installing *source* package ‘odbc’ ...
** package ‘odbc’ successfully unpacked and MD5 sums checked
** using staged installation
Found odbc_config cflags and libs!
PKG_CFLAGS=-DHAVE_UNISTD_H -DHAVE_PWD_H -DHAVE_SYS_TYPES_H -DHAVE_LONG_LONG -DSIZEOF_LONG_INT=8 -I/opt/R/x86_64/include
PKG_LIBS=-L/opt/R/x86_64/lib -lodbc
** libs
using C++ compiler: ‘Homebrew clang version 14.0.6’
using SDK: ‘MacOSX14.0.sdk’

<snip>

> DBI::dbConnect(odbc::odbc(), dsn="snowflake_db", uid="<snip>", pwd="<snip>")
<OdbcConnection> <snip>@Snowflake
  Database: SNOWFLAKE_SAMPLE_DATA
  Snowflake Version: 7.43.0
```

**Comment by hadley** [comment]
Hmmm, I have 15 and it doesn't crash for me.

```
$ clang -v
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: arm64-apple-darwin23.1.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by detule** [comment]
That's a good data-point.  I think issue may be isolated to `x86_64`

**Comment by gaborcsardi** [comment]
I can reproduce this with the MS SQL Server driver, with:
```
❯ clang --version
Apple clang version 15.0.0 (clang-1500.1.0.2.5)
Target: x86_64-apple-darwin22.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

❯ sw_vers
ProductName:		macOS
ProductVersion:		13.6.1
BuildVersion:		22G313
```

**Comment by gaborcsardi** [comment]
Some random notes, will extend, possibly:

* Also crashes with CRAN's static unixodbc from https://mac.r-project.org/bin/
* Also crashes on R-devel
* It does not crash with `CXXFLAGS+=-O0`

**Comment by gaborcsardi** [comment]
See the [linked PR](https://github.com/r-dbi/odbc/pull/677) for why this is happening. Hopefully the PR will help you implement a proper and complete fix.

**Comment by detule** [comment]
Fixed in #677
